### ISSUE TITLE
Parameter tuning for stacked learner

### ISSUE THREAD

**Issue by njbernstein** [mlr-org/mlr#697]
Hi,

This isn't an issue an so much I can't figure out how to tune a stacked learner. Here's my attempt:

```
## create mlr task and convert factors to dummy features
trainTask = makeRegrTask(data = train, target = "Response")
trainTask = createDummyFeatures(trainTask)
testTask = makeRegrTask(data = test, target = "Response")
testTask = createDummyFeatures(testTask)


## create mlr learner
set.seed(123)
lrn = makeLearner("regr.xgboost")
lrn$par.vals = list(
  nthread             = 5,
  nrounds             = 10,
  print.every.n       = 50,
  objective           = "reg:linear"
)
# missing values will be imputed by their median
lrn = makeImputeWrapper(lrn, classes = list(numeric = imputeMedian(), integer = imputeMode()))

lrn2 <- makeLearner("regr.avNNet")
lrn2 = makeImputeWrapper(lrn2, classes = list(numeric = imputeMedian(), integer = imputeMode()))

lrn2 <- makeLearner("regr.avNNet")
lrn2 = makeImputeWrapper(lrn2, classes = list(numeric = imputeMedian(), integer = imputeMode()))

## Create Evaluation Function
SQWKfun = function(x = seq(1.5, 7.5, by = 1), pred) {
  preds = pred$data$response
  true = pred$data$truth 
  cuts = c(min(preds), x[1], x[2], x[3], x[4], x[5], x[6], x[7], max(preds))
  preds = as.numeric(Hmisc::cut2(preds, cuts))
  err = Metrics::ScoreQuadraticWeightedKappa(preds, true, 1, 8)
  return(-err)
}
SQWK = makeMeasure(id = "SQWK", minimize = FALSE, properties = c("regr"), best = 1, worst = 0,
                   fun = function(task, model, pred, feats, extra.args) {
                     return(-SQWKfun(x = seq(1.5, 7.5, by = 1), pred))
                   })

# Do it in parallel with parallelMap
library(parallelMap)
parallelStartSocket(3)
parallelExport("SQWK", "SQWKfun", "trainLearner.regr.xgboost", "predictLearner.regr.xgboost" , "makeRLearner.regr.xgboost")
## This is how you could do hyperparameter tuning
# # 1) Define the set of parameters you want to tune (here 'eta')



lrn2 <- makeLearner("regr.avNNet")
lrn2$par.vals = list(
  maxit             = 100,
  size =5
)

lrn3 <- makeLearner("regr.frbs") 
 
stacked.lrn <- makeStackedLearner(base.learners = list(lrn2,lrn3), super.learner = lrn , method = "stack.cv" ,predict.type = "response")
stacked.lrn <- makeImputeWrapper(stacked.lrn, classes = list(numeric = imputeMedian(), integer = imputeMode()))

# perform crossvalidation in parallel
ps = makeParamSet(
  makeNumericParam("eta", lower = 0.01, upper = 0.3),
  makeIntegerParam("max_depth", lower =  2, upper = 12),
  makeIntegerParam("min_child_weight", lower =  1, upper = 500),
  makeNumericParam("gamma",lower=1,upper=100),
  makeNumericParam("colsample_bytree", lower = 1, upper = 2, trafo = function(x) x/2),
  makeNumericParam("subsample", lower = 1, upper = 2, trafo = function(x) x/2)
)
# # 2) Use 3-fold Cross-Validation to measure improvements
rdesc = makeResampleDesc("CV", iters = 2L)
# # 3) Here we use Random Search (with 10 Iterations) to find the optimal hyperparameter
ctrl =  makeTuneControlRandom(maxit = 10)
# # 4) now use the learner on the training Task with the 3-fold CV to optimize your set of parameters and evaluate it with SQWK
res = tuneParams(stacked.lrn, task = trainTask, resampling = rdesc, par.set = ps, control = ctrl, measures = SQWK)
res
# # 5) set the optimal hyperparameter
stacked.lrn = setHyperPars(stacked.lrn, par.vals = res$x)


cv = crossval(stacked.lrn, trainTask, iter = 2, measures = SQWK, show.info = TRUE)

#bag.lrn <- makeBaggingWrapper(lrn,  bw.iters = 100, bw.feats = 0.5 )
#cv = crossval(bag.lrn, trainTask, iter = 3, measures = SQWK, show.info = TRUE)

parallelStop()
## now try to find the optimal cutpoints that maximises the SQWK measure based on the Cross-Validated predictions
optCuts = optim(seq(1.5, 7.5, by = 1), SQWKfun, pred = cv$pred)
optCuts

## now train the model on all training data
tr = train(stacked.lrn, trainTask)

## predict using the optimal cut-points 
pred = predict(tr, testTask)
preds = as.numeric(Hmisc::cut2(pred$data$response, c(-Inf, optCuts$par, Inf)))
table(preds)
```

I get the error at the tuning step
```
res = tuneParams(stacked.lrn, task = trainTask, resampling = rdesc, par.set = ps, control = ctrl, measures = SQWK)
Error in checkTunerParset(learner, par.set, measures, control) : 
  Can only tune parameters for which learner parameters exist: eta,max_depth,min_child_weight,gamma,colsample_bytree,subsample
```

Any help with this much appreciated. I'm new to this stuff as well, so any pointers would be great

**Comment by larskotthoff** [comment]
This is because stacked learners don't support tuning of the parameters of the super learner at the moment. It looks like that shouldn't be too hard to add though.

**Comment by njbernstein** [comment]
Great! Just let me know if there is anything I can do to help (e.g. testing or anything)

**Comment by larskotthoff** [comment]
I've had a look at this and it seems that we'll need a new class for StackedLearners extending BaseEnsemble. @berndbischl Do you see a different way?

**Comment by giuseppec** [comment]
Hi, 

what you could do in the mean-time to get this working (I think there is 1 week left for your kaggle-competition) is the following:

```r
base.learners = list(lrn2, lrn3) 
# create the cross-validated predictions using the base learners
cvList = sapply(base.learners, function(x) crossval(x, trainTask) )
# order cv-predictions
for(i in 1:length(cvList))
  cvList[[i]]$pred$data = cvList[[i]]$pred$data[order(cvList[[i]]$pred$data$id), ]
# create a new data set that uses the cv-predictions
train2 = cbind(getTaskData(trainTask), cvList) # you can use the original features or just use the cv-predictions
trainTask2 = makeRegrTask(data = train2, target = "Response")
```
Then you can create your super learner and tune this on the `trainTask2`. Hope this helps (maybe there is a typo in the code above, I did not test it). 

Maybe we should split the "training of base learners" and "training of the super learner" in two independent "parts", currently this is done in one single function. I can talk with @berndbischl how to do this properly.

PS: good luck on kaggle with this ;)

**Comment by njbernstein** [comment]
Thanks so much for the help. Thats a nice elegant solution to what I wanted to accomplish. I'll do my best on Kaggle thanks again for your help

**Comment by berndbischl** [comment]
Ok, 2 general things: 

1) I have on person looking in depth at stacking during his masters. So we can hopefully expect more progress here in general. 
(Hi Phillip)

2) I checked the stacking code. We actually support a lot of tuning already. 
Tuning of all baselearner params already works, via inheritance from BaseEnsemble (actually I kinda forgot that I did when I looked first....)

Like Lars said it does not work for the superlearner. We should open a clean issue for this and Phillip / me could do this soon.

OK?
